runtime/planificator.js

```diff
@@ -6,13 +6,84 @@
 // http://polymer.github.io/PATENTS.txt
 
 import {assert} from '../platform/assert-web.js';
+import {now} from '../platform/date-web.js';
 import {InitSearch} from './strategies/init-search.js';
 import {Planner} from './planner.js';
 import {Speculator} from './speculator.js';
 import {SuggestionComposer} from './suggestion-composer.js';
 
 let defaultTimeoutMs = 5000;
 
+class ReplanQueue {
+  constructor(planificator, options) {
+    this._planificator = planificator;
+    this._options = options || {};
+    this._options.defaultReplanDelayMs = this._options.defaultReplanDelayMs || 3000;
+
+    this._changes = [];
+    this._replanTimer = null;
+    this._planificator.registerStateChangedCallback(this._onPlanningStateChanged.bind(this));
+  }
+  addChange() {
+    this._changes.push(now());
+    if (this._isReplanningScheduled()) {
+      this._postponeReplan();
+    } else if (!this._planificator.isPlanning) {
+      this._scheduleReplan(this._options.defaultReplanDelayMs);
+    }
+  }
+
+  _onPlanningStateChanged(isPlanning) {
+    if (isPlanning) {
+      // Cancel scheduled planning.
+      this._cancelReplanIfScheduled();
+      this._changes = [];
+    } else if (this._changes.length > 0) {
+      // Schedule delayed planning.
+      let timeNow = now();
+      this._changes.forEach((ch, i) => this._changes[i] = timeNow);
+      this._scheduleReplan(this._options.defaultReplanDelayMs);
+    }
+  }
+  _isReplanningScheduled() {
+    return Boolean(this._replanTimer);
+  }
+  _scheduleReplan(intervalMs) {
+    this._cancelReplanIfScheduled();
+    this._replanTimer = setTimeout(() => this._planificator._requestPlanning(), intervalMs);
+  }
+  _cancelReplanIfScheduled() {
+    if (this._isReplanningScheduled()) {
+      clearTimeout(this._replanTimer);
+      this._replanTimer = null;
+    }
+  }
+  _postponeReplan() {
+    if (this._changes.length <= 1) {
+      return;
+    }
+    let now = this._changes[this._changes.length - 1];
```

Comment:
I don't think we need to track all of `this._changes`, we need to know when the current replan was first triggered, and the time of the most recent invalidation. But it's OK to leave it like this too.

Reply:
I kept it because I was planning to actually track what data has changed and replan depending on it. 
If I end up not using it, I'll remove it in the iterations.